File: day-03/python/part2.py
#!/usr/bin/env python3

import sys

data = [x.strip() for x in open(sys.argv[1], 'r').readlines()]

slopes = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]  # right, down
# right = [2, 3, 4, 6, 8, 9, 12, 16, 18, 24, 32, 36, 48, 54, 64]
# down = [1, 5, 7, 11, 13, 17, 19, 23, 25, 29, 31, 35, 37, 41, 47]
# slopes = [(x,y) for x in right for y in down]

numRows = len(data)  # changes between the example and actual input
numCols = len(data[0])  # ^

# m = y2 - y1 / x2 - x1 -> basically y/x

totalTrees = 1

def solve(slope: tuple) -> int:
    trees = 0
    x = 0
    for y in range(0, numRows, slope[1]):  # -> step right by factor of right 
        if data[y][x] == '#':  # this is not a 2d array, it is checking the y and x-axis for a "tree"
            trees += 1
        x = (x + slope[0]) % numCols  # allow for "looping" around on the x-axis
    return trees

for s in slopes:
    totalTrees *= solve(s)  # pass each set of slopes into solve and mul the results 

print(totalTrees)

File: day-03/lua/part2.lua
#!/usr/bin/env lua5.4

local function parse(inp)
    local num = {}
    for str in io.lines(inp) do
        num[#num+1] = str
    end
    return num
end

local data = parse(arg[1])
local slopes = {{1, 1}, {3, 1}, {5, 1}, {7, 1}, {1, 2}}

local numRows = #data
local numCols = string.len(data[1])

local totalTrees = 1

local function solve(slope)
    local trees = 0
    local x = 0
    for y=1,numRows,slope[2] do
        if data[y]:sub(x+1, x+1) == '#' then
            trees = trees + 1
        end
        x = (x + slope[1]) % numCols
    end
    return trees
end

for i=1, #slopes do
    totalTrees = totalTrees * solve(slopes[i])
end

print("total trees: ".. totalTrees)
